homeassistant/components/calendar/trigger.py

```diff
@@ -87,18 +90,17 @@ async def _fetch_events(self, now: datetime.datetime) -> None:
         # returned events may have already started but matched the start/end time
         # filtering above, so exclude any events that have already passed the
         # trigger time.
-        event_list = [
-            (dt_util.as_utc(event.start_datetime_local), event) for event in events
-        ]
+        event_list = []
+        for event in events:
+            event_time = (
+                event.start_datetime_local
+                if self._event_type == EVENT_START
+                else event.end_datetime_local
+            )
+            if event_time > now:
+                event_list.append((event_time, event))
         event_list.sort(key=lambda x: x[0])
-
-        self._events.extend(
```

Comment:
I've been going back and forth about how to best make this act like a cursor, even though its not.  This API doesn't have way to page or iterate that is "correct", to ensure we don't get the same events multiple times.

With end events this gets a little more complex than start events, i think.  This likely gets event more complex with offsets.

The strategy now here is to use the alarm time `now` as the way to page, and try not to create any gaps (e.g. use the alarm time as the next time). 

The apis use times as exclusive though, so we kind of have to assume we won't run into an event on the /exact/ boundary of an event here.

Looking for feedback.

Reply:
On line 84 you call it `start_date`. That should be named `start_datetime` as it's a datetime. 

If it's exclusive, you could set start time to "last_endtime - 1 second" to overlap it I guess? However, treating it as a cursor assumes that events can't be added/removed ?

Reply:
Yeah, it is exclusive, and i was considering something like that (-1), just felt bad about. Let me make this change in a follow up PR and also spell this out more clearly in the developer docs.

As far as picking up new events: Yeah, true, the advantage of this approach now is that it can kind of look backwards within the window.